src/globus_cli/principal_resolver.py

```diff
@@ -0,0 +1,98 @@
+"""
+A wrapper which nicely integrates SDK IdentityMaps into the CLI.
+
+This was originally implemented in gcs-cli and ported into here.
+
+A "resolver" is defined with a field it uses in each item from the response data as the
+full principal value. This is the "key".
+Typical usage should be to define a resolver in each command which needs resolution, to
+provide the key to use. Definition may have to happen outside of a command to
+accommodate field lists which are defined as constants, and this is supported.
+
+A resolver provides a field handler for principal URNs + a pagination callback to get
+all principals from each page of results added to the ID map.
+It can also be initialized to take identity IDs (not in Principal URN format).
+
+This lets us get some of the benefit of the IdentityMap bulk calls without forcing the
+whole paginated call to be walked at once.
+
+This also lets us keep the resolution work only in the text-mode printed output (and not
+applied on JSON output).
+"""
+from globus_sdk import IdentityMap
+
+from globus_cli.services.auth import get_auth_client
+
+IDENTITY_URN_PREFIX = "urn:globus:auth:identity:"
+
+
+class PrincipalResolver:
+    """
+    Everything is done lazily via properties so that nothing happens during
+    start up.
+
+    Pass ``PrincipalResolver.field`` as a field key for output printing
+
+    Pass ``PrincipalResolver.page_callback`` as a page callback for output printing.
+
+    Usage:
+
+    >>> PrincipalResolver("urnfield")
+
+    creates a resolver which pulls principal URNs from the field named "urnfield"
+
+    >>> PrincipalResolver("idfield", use_urns=False)
+
+    creates a resolver which pulls Identity IDs from the field named "idfield"
+    """
+
+    def __init__(self, key, use_urns=True):
+        self.key = key
+        self.use_urns = use_urns
+        self._idmap = None
+
+    @property
+    def idmap(self):
+        if not self._idmap:
+            self._idmap = IdentityMap(get_auth_client())
+        return self._idmap
+
+    def _raw_id_from_object(self, obj):
+        """
+        returns a 3-tuple, (resolved?, original, value)
+        so that callers know if this helper thinks it got a good result or not
+        """
+        value = obj[self.key]
+        # if not using URNs, the "raw ID" is just the value and it "always works"
+        if not self.use_urns:
+            return (True, value, value)
+
+        # otherwise, check
+        # if it doesn't have the URN prefix, it is not a valid URN, so this lookup
+        # failed -- the result is (False, ...) to indicate failure
+        if not value.startswith(IDENTITY_URN_PREFIX):
+            return (False, value, value)
```

Comment:
I wonder if just `raise ValueError` in this case would be a little more idiomatic/easier to parse?

Reply:
Maybe... I think I'd use a special subclass of ValueError for this though, so we don't accidentally catch anything else.

Note that it's not propagating anywhere, since we'd catch it immediately. I'll take a look at changing it in a few minutes.

Reply:
Picking it apart, the reason for this oddness was to pass `value` back regardless of whether or not this succeeds.

I added `InvalidPrincipalError(ValueError)` and attached the value seen there instead. The result is marginally simpler, so I'll leave the refactor in place.